crates/ide_assists/src/handlers/add_turbo_fish.rs

```diff
@@ -339,4 +345,42 @@ fn main() {
             "Add `: _` before assignment operator",
         );
     }
+
+    #[test]
+    fn add_turbo_fish_function_lifetime_parameter() {
+        check_assist(
+            add_turbo_fish,
+            r#"
+fn make<'a, T, A>(t: T, a: A) {}
+fn main() {
+    make$0(5, 2);
+}
+"#,
+            r#"
+fn make<'a, T, A>(t: T, a: A) {}
+fn main() {
+    make::<${0:_,_}>(5, 2);
+}
+"#,
+        );
+    }
+
+    #[test]
+    fn add_turbo_fish_function_const_parameter() {
+        check_assist(
+            add_turbo_fish,
+            r#"
+fn make<T, const N: usize>(t: T) {}
+fn main() {
+    make$0(3);
+}
+"#,
+            r#"
+fn make<T, const N: usize>(t: T) {}
+fn main() {
+    make::<${0:_,_}>(3);
```

Comment:
Also, the snippet here(and on line `362`) are actually wrong(this is an oversight from your previous PR), this should be `make::<${1:_},${2:_}>(3);`

Reply:
or rather `make::<${1:_},${0:_}>(3);`(since the order goes `1, 2, 3, ..., n - 1, n, 0`, basically 0 is always the last tabstop)

Reply:
Not saying you have to fix that in this PR btw, forgot to mention that. Just wanted to add that as a side note(though you are are welcome to fix this as well if you have the time)

bors d+

Reply:
I don't mind fixing this now, I just need a little bit more information on what I'm doing here. What do the `1` and `0` represent in your example of `make::<${1:_},${0:_}>(3);`?

I presume I'd have to change it here: https://github.com/rust-analyzer/rust-analyzer/blob/2fb6f5e46a5f53a78c2f750dd3c8e79dd418c4c9/crates/ide_assists/src/handlers/add_turbo_fish.rs#L89
What about the `None` case underneath? Does that exclude the `1` and `0` bits?

Reply:
So the $n things are tabstops, which define an order of where the cursor in the client can jump to(usually by pressing tab), `{$n:_}` is a special form of this that just defines a placeholder/template text(the `_` part is the placeholder text). The numbers define the order, where as 0 is always the last tab stop that gets jumpted to, otherwise 1 is the first, 2 is the second etc.

The None case below is for when the client doesnt support snippets(tabstops), so there we don't want any of the $ shenanigans.

So for the some branch we would want to see this `<${1:_},${2:_},${0:_}>` if we had 2 tabstops and for the None branch it should be `<_, _, _>`, if we had 3 generics as an example.

Reply:
Thanks for the explanation! I've pushed a commit that sorts out these tabstops. The code is a bit wordy though so perhaps you might see an improvement I can make here?

It looks like the unit tests always test the `Some()` branch -- is there a way to turn snippet support off in the tests? That way I could write a test for the other case as well.

Reply:
> It looks like the unit tests always test the Some() branch -- is there a way to turn snippet support off in the tests? That way I could write a test for the other case as well.

I don't think so, that is a good point that we should adress at some point though.